File: src/main/java/com/group/libraryapp/service/url/UrlShortenerService.java
package com.group.libraryapp.service.url;

import com.group.libraryapp.domain.url.Url;
import com.group.libraryapp.domain.url.UrlRepository;
import com.group.libraryapp.dto.url.request.ShortenUrlRequest;
import com.group.libraryapp.dto.url.response.ShortenUrlResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.security.SecureRandom;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@Service
public class UrlShortenerService {

    private static final Logger logger = LoggerFactory.getLogger(UrlShortenerService.class);
    private static final String CHARACTERS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    private static final int KEY_LENGTH = 6;
    
    // In-memory hashmap for storing shortened URLs
    private final Map<String, String> urlMap = new HashMap<>();
    private final UrlRepository urlRepository;
    private final SecureRandom random = new SecureRandom();

    public UrlShortenerService(UrlRepository urlRepository) {
        this.urlRepository = urlRepository;
    }

    @Transactional
    public ShortenUrlResponse shortenUrl(ShortenUrlRequest request) {
        String originalUrl = request.getUrl();
        logger.info("Shortening URL: {}", originalUrl);

        String shortKey = generateUniqueKey();
        
        // Store in both in-memory map and database
        urlMap.put(shortKey, originalUrl);
        urlRepository.save(new Url(shortKey, originalUrl));
        
        logger.info("URL shortened successfully. Short key: {}", shortKey);
        return new ShortenUrlResponse(shortKey);
    }

    @Transactional(readOnly = true)
    public String getOriginalUrl(String shortKey) {
        logger.info("Looking up original URL for short key: {}", shortKey);
        
        // First check in-memory map
        String originalUrl = urlMap.get(shortKey);
        
        // If not found in memory, try the database
        if (originalUrl == null) {
            Optional<Url> urlOptional = urlRepository.findByShortKey(shortKey);
            if (urlOptional.isPresent()) {
                originalUrl = urlOptional.get().getOriginalUrl();
                // Cache in memory for future lookups
                urlMap.put(shortKey, originalUrl);
            } else {
                logger.info("No URL found for key: {}", shortKey);
                return null;
            }
        }
        
        logger.info("Original URL found: {}", originalUrl);
        return originalUrl;
    }

    private String generateUniqueKey() {
        String key;
        do {
            key = generateKey();
        } while (urlMap.containsKey(key) || urlRepository.existsByShortKey(key));
        
        return key;
    }

    private String generateKey() {
        StringBuilder sb = new StringBuilder(KEY_LENGTH);
        
        // Ensure first character is a letter
        sb.append(CHARACTERS.charAt(random.nextInt(52))); // First 52 chars are letters
        
        // Add remaining characters (can be letters or digits)
        for (int i = 1; i < KEY_LENGTH; i++) {
            sb.append(CHARACTERS.charAt(random.nextInt(CHARACTERS.length())));
        }
        
        return sb.toString();
    }
}

File: src/main/kotlin/com/griotold/mcp_demo/service/UrlShortenerService.kt
package com.griotold.mcp_demo.service

import com.griotold.mcp_demo.domain.Url
import com.griotold.mcp_demo.dto.ShortenUrlRequest
import com.griotold.mcp_demo.dto.ShortenUrlResponse
import com.griotold.mcp_demo.repository.UrlRepository
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.security.SecureRandom

@Service
class UrlShortenerService(
    private val urlRepository: UrlRepository
) {
    private val logger = KotlinLogging.logger {}
    private val urlMap: MutableMap<String, String> = HashMap()
    private val random = SecureRandom()
    
    companion object {
        private const val CHARACTERS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        private const val KEY_LENGTH = 6
    }
    
    @Transactional
    fun shortenUrl(request: ShortenUrlRequest): ShortenUrlResponse {
        val originalUrl = request.url
        logger.info { "Shortening URL: $originalUrl" }
        
        val shortKey = generateUniqueKey()
        
        // Store in both in-memory map and database
        urlMap[shortKey] = originalUrl
        urlRepository.save(Url(shortKey, originalUrl))
        
        logger.info { "URL shortened successfully. Short key: $shortKey" }
        return ShortenUrlResponse(shortKey)
    }
    
    @Transactional(readOnly = true)
    fun getOriginalUrl(shortKey: String): String? {
        logger.info { "Looking up original URL for short key: $shortKey" }
        
        // First check in-memory map
        var originalUrl = urlMap[shortKey]
        
        // If not found in memory, try the database
        if (originalUrl == null) {
            val urlOptional = urlRepository.findByShortKey(shortKey)
            if (urlOptional.isPresent) {
                originalUrl = urlOptional.get().originalUrl
                // Cache in memory for future lookups
                urlMap[shortKey] = originalUrl
            } else {
                logger.info { "No URL found for key: $shortKey" }
                return null
            }
        }
        
        logger.info { "Original URL found: $originalUrl" }
        return originalUrl
    }
    
    private fun generateUniqueKey(): String {
        var key: String
        do {
            key = generateKey()
        } while (urlMap.containsKey(key) || urlRepository.existsByShortKey(key))
        
        return key
    }
    
    private fun generateKey(): String {
        val sb = StringBuilder(KEY_LENGTH)
        
        // Ensure first character is a letter
        sb.append(CHARACTERS[random.nextInt(52)]) // First 52 chars are letters
        
        // Add remaining characters (can be letters or digits)
        for (i in 1 until KEY_LENGTH) {
            sb.append(CHARACTERS[random.nextInt(CHARACTERS.length)])
        }
        
        return sb.toString()
    }
}
